modules/broker-core/src/main/java/io/ballerina/messaging/broker/core/transaction/Registry.java

```diff
@@ -140,4 +181,34 @@ public void forget(Xid xid) throws ValidationException {
             unregister(xid);
         }
     }
+
+    public void setTimeout(Xid xid, long timeout, TimeUnit timeUnit) throws ValidationException {
+        Branch branch = branchMap.get(xid);
+        if (Objects.isNull(branch)) {
+            throw new ValidationException(UNKNOWN_XID_ERROR_MSG + xid);
+        }
+
+        if (timeout == 0) {
+            return;
+        }
+
+        ScheduledFuture<?> future = branchTimeoutExecutorService.schedule(() -> {
+
+            LOGGER.debug("timing out dtx task with xid {}", xid);
+            synchronized (branch) {
+                if (branch.getState() == Branch.State.PREPARED) {
+                    LOGGER.debug("Branch already prepared. Won't be timed out. Xid {}", xid);
+                    return;
+                }
+                try {
+                    rollback(xid);
+                    branch.setState(Branch.State.TIMED_OUT);
```

Comment:
Will it be safer to set the state first?

Reply:
Idea was to set the branch to TIMEDOUT state if the branch rollback was successful. If it failed It won't get timed out.